ui/app/AppLayouts/Shell/ShellAdaptor.qml

```diff
@@ -0,0 +1,401 @@
+    }
+
+    ObjectProxyModel {
+        id: dappsModel
+
+        sourceModel: root.dappsBaseModel
+        delegate: QtObject {
+            readonly property string key: Constants.appSection.dApp + ';' + model.url
+            readonly property string id: model.url
+            readonly property string name: model.name || StringUtils.extractDomainFromLink(model.url)
+            readonly property string icon: model.iconUrl || "dapp"
+            readonly property color color: Theme.palette.primaryColor1
+
+            readonly property url connectorBadge: model.connectorBadge
+        }
+        expectedRoles: ["url", "name", "iconUrl", "connectorBadge"]
+        exposedRoles: ["key", "id", "name", "icon", "color", "connectorBadge"]
+    }
+
+    ConcatModel {
+        id: combinedModel
+        sources: [
+            SourceModel {
+                model: root.showCommunities ? communitiesModel : null
+                markerRoleValue: Constants.appSection.community
+            },
+            SourceModel {
+                model: root.showWallets ? walletsModel : null
+                markerRoleValue: Constants.appSection.wallet
+            },
+            SourceModel {
+                model: root.showSettings ? settingsModel : null
+                markerRoleValue: Constants.appSection.profile
+            },
+            SourceModel {
+                model: root.showChats ? chatsModel : null
+                markerRoleValue: Constants.appSection.chat
+            },
+            SourceModel {
+                model: root.showDapps ? dappsModel : null
+                markerRoleValue: Constants.appSection.dApp
+            }
+        ]
+
+        markerRoleName: "sectionType"
+        expectedRoles: ["key", "id", "enabled", "name", "icon", "color", "hasNotification", "notificationsCount", // common props
+            "banner", "members", "activeMembers", "pending", "banned", // community
+            "isExperimental", // settings
+            "walletType", "currencyBalance", // wallet
+            "connectorBadge" // dapp
+        ]
+    }
+
+    Settings {
+        id: shellSettings
+        category: "Shell_%1".arg(root.profileId)
+    }
+
+    ObjectProxyModel { // provides a writable overlay for "timestamp" and "pinned" roles
+        id: shellProxyModel
+
+        sourceModel: combinedModel
+        delegate: QtObject {
+            property real timestamp
+            property bool pinned
+        }
+
+        exposedRoles: ["timestamp", "pinned"]
+    }
+
+    function setPinned(key, pinned) {
+        const idx = ModelUtils.indexOf(shellProxyModel, "key", key)
+        if (idx > -1) {
+            shellProxyModel.proxyObject(idx).pinned = pinned
+        }
+    }
+
+    function setTimestamp(key, timestamp) {
+        const idx = ModelUtils.indexOf(shellProxyModel, "key", key)
+        if (idx > -1) {
+            shellProxyModel.proxyObject(idx).timestamp = timestamp
+        }
+    }
+
+    function save() {
```

Comment:
instead saving only onDestruciton we could also use `Timer` reseted by successful `setPinned` or `setTimestamp`. Then there is no risk of loosing the arrangement e.g. in case of app crash

Reply:
I tried forcefully killing the app and it still saved the entries :)